scipy/stats/_binned_statistic.py

```diff
@@ -513,8 +517,9 @@ def binned_statistic_dd(sample, values, statistic='mean',
     if not callable(statistic) and statistic not in known_stats:
         raise ValueError('invalid statistic %r' % (statistic,))
 
-    if not np.isfinite(values).all() or not np.isfinite(sample).all:
-        raise ValueError('%r or %r contains non-finite values.' % (sample, values,))
+    # NOTE: for _bin_edges(), see e.g. gh-11365
+    if isinstance(bins, int) and not np.isfinite(sample).all():
```

Comment:
Taking another look, I just noticed this: `isinstance(bins, int)` will not recognize numpy integers, e.g. `np.int64(10)`.  Lately we have settled on checking for integerness by passing the value through `operator.index`.   That is, if `operator.index(bins)` doesn't raise a `TypeError`, then `bins` is an integer.  So at a minimum, we could add this before the current line:
```
    try:
        bins = operator.index(bins)
    except TypeError:
        # bins is not an integer
        pass
    # If bins was an integer-like object, now it is an actual Python int.
```
Then the test `isinstance(bins, int)` will work.  There is probably a prettier way to do that, but it would work.

Reply:
> @rlucasy, finally got back to this. Looks good.
> 
> The master branch now has a couple lines added to the unit test to avoid making the inplace change of `self.u` visible after the test is finished. That could be simplified by using `u = self.u.copy()`, and then use the code in this PR for the rest of the test. Or do the swapping that is in master--either is fine.

I finally got a few spare moments to look at this PR again, sorry it's taken me awhile. I rebased and resolved the conflicts as you suggest. However, the test you are referencing should be removed as it was incorrect to throw the error to begin with. I'll do a force push today with the resolution of this and the other issue you noted.

Reply:
> Taking another look, I just noticed this: `isinstance(bins, int)` will not recognize numpy integers, e.g. `np.int64(10)`. Lately we have settled on checking for integerness by passing the value through `operator.index`. That is, if `operator.index(bins)` doesn't raise a `TypeError`, then `bins` is an integer. So at a minimum, we could add this before the current line:
> 
> ```
>     try:
>         bins = operator.index(bins)
>     except TypeError:
>         # bins is not an integer
>         pass
>     # If bins was an integer-like object, now it is an actual Python int.
> ```
> 
> Then the test `isinstance(bins, int)` will work. There is probably a prettier way to do that, but it would work.

Thanks for catching this Warren, always something with the types it seems :) 
I'll add a test for this and the suggested fix you've noted here. There maybe a prettier way to do it but you're in good company with your suggestion. see e.g.: https://stackoverflow.com/a/54381905/3164100 so I see no reason to try something fancy.